Am/excessive gh config validation reqs

Apmats opened the pull request:
## Closes https://github.com/elastic/connectors-py/issues/###


<!--Provide a general description of the code changes in your pull request.
If the change relates to a specific issue, include the link at the top.

If this is an ad-hoc/trivial change and does not have a corresponding
issue, please describe your changes in enough details, so that reviewers
and other team members can understand the reasoning behind the pull request.-->

## Checklists

<!--You can remove unrelated items from checklists below and/or add new
items that may help during the review.-->

#### Pre-Review Checklist
- [ ] this PR does NOT contain credentials of any kind, such as API keys or username/passwords (double check `config.yml.example`)
- [ ] this PR has a meaningful title
- [ ] this PR links to all relevant github issues that it fixes or partially addresses
- [ ] if there is no GH issue, please create it. Each PR should have a link to an issue
- [ ] this PR has a thorough description
- [ ] Covered the changes with automated tests
- [ ] Tested the changes locally
- [ ] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] For bugfixes: backport safely to all minor branches still receiving patch releases
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

#### Changes Requiring Extra Attention

<!--Please call out any changes that require special attention from the
reviewers and/or increase the risk to availability or security of the
system after deployment. Remove the ones that don't apply.-->

- [ ] Security-related changes (encryption, TLS, SSRF, etc)
- [ ] New external service dependencies added.

## Related Pull Requests

<!--List any relevant PRs here or remove the section if this is a standalone PR.

* https://github.com/elastic/.../pull/123-->

## Release Note

<!--If you think this enhancement/fix should be included in the release notes,
please write a concise user-facing description of the change here.
You should also label the PR with `release_note` so the release notes
author(s) can easily look it up.-->

lorenabalan commented on connectors/sources/github.py:
```diff
@@ -197,6 +197,11 @@ class GithubQuery(Enum):
     }
     }
     """
+    BATCH_REPO_QUERY_TEMPLATE = """
```
I don't see this being used though. Should it be in L1103?

Apmats commented on connectors/sources/github.py:
```diff
@@ -1002,6 +1036,98 @@ async def get_foreign_repo(self, repo_name):
+
+        # Build variable declarations for the query
+        variable_declarations = []
+        for var_name in variables.keys():
```
Done!

Apmats commented on connectors/sources/github.py:
```diff
@@ -807,13 +813,36 @@ async def graphql(self, query, variables=None):
+                # All errors are ignored, return just the data part without errors
+                return exception.response.get("data", {})
+
+            if ignore_errors:
```
Yeah this was already me doing a first pass refactoring of the extremely verbose code genai gave me. I did what you said pretty much now.

Apmats commented:
Addressed review comments from @lorenabalan and also added some typing. Another round of manual testing and a self-review pending - I think I can for sure add some more types.

Apmats commented on connectors/sources/github.py:
```diff
@@ -1002,6 +1028,93 @@ async def get_foreign_repo(self, repo_name):
+        """
+        results = {}
+
+        # Process repositories in batches to avoid hitting GraphQL complexity limits
```
I think I hit an alias limit while testing this out, but I can't reproduce now. I might be going crazy.
 
Complexity limits I think is about how many nodes you can reach in a single query - currently they are here https://docs.github.com/en/enterprise-cloud@latest/graphql/overview/rate-limits-and-query-limits-for-the-graphql-api#node-limit - I don't understand fully though and I don't know how much of a subject to change they are. 

Hitting them would result in an error that would get raised here and captured in the connector config error I believe (since we only handle rate limit errors and retry after waiting), but I think with the current query structure (which doesn't have nesting) it would require you to have configured a pathologically large number of repos to sync.
But having batching already built in helps prevent any such pathologic case or any change to the downstream query that adds to that complexity.

Another problem with a big number of handling a big number of repos here is the fact that we would hit timeouts probably consistently:
https://docs.github.com/en/enterprise-cloud@latest/graphql/overview/rate-limits-and-query-limits-for-the-graphql-api#timeouts
Testing with 300 repos seems to take me close to timeout territory.

So a) batching will not come into play in realistic scenarios (I can increase to like 50, and most users won't be syncing against more than 50 repos I would guess) and b) if it does come into play it might help us avoid some errors.

What are your thoughts?

Apmats commented on connectors/sources/github.py:
```diff
@@ -807,13 +822,24 @@ async def graphql(self, query, variables=None):
-            msg = f"Error while executing query. Exception: {exception.response.get('errors')}"
+
+            if all(error.get("type") in ignore_errors for error in errors):
+                # All errors are ignored, return just the data part without errors
+                return exception.response.get("data", {})
```
Given you're passing the errors you don't really consider errors for your case, hitting them is within the happy path 🤔 .

Eg. when this is called like in this PR to check for multiple repos' existence.
OR
if you use the same method to verify repo doesn't exist and thus was indeed deleted, again, part of the happy path in that case.

artem-shelkovnikov replied:
I'd look at it from the perspective of SDH or a customer who doesn't look into the connector code base at all.

We roll something out, then there comes the question of "hmmmm, something went wrong - I don't get the data, lemme look at logs" and logs should answer this.

In the context of this PR I want to understand whether we log sufficient amount of info to understand whether the repository is there or not (I don't think the `errors` parameter is passed elsewhere, right?). Now thinking of it, we likely do, because we log a warning saying "repos A, B and C are not available".

If, however, more methods in code start using the method, for debugging/SDH/troubleshooting purpose this can be useful?

So in the end yeah, likely no need to log anything here _yet_.

Apmats replied:
In this case, this always results in an error being raised when we ignore this errors within the context of a GraphQL query, so yes we get visibility.

I don't know why I had to debate this, I'll add the log line. If this ends up generating too many logs it's your/our fault for running in debug mode in a prod-like environment, turn it off! 😅

artem-shelkovnikov replied:
I do agree with the debate though, we will end up reading debug logs.

Apmats commented on connectors/sources/github.py:
```diff
@@ -1586,6 +1743,9 @@ async def _fetch_installations(self):
         return self._installations
 
     async def _get_personal_repos(self, user):
```
This path is still valid because we support wildcards - we don't know the repos to fetch, we fetch all-repos-for-user here for example.

We could probably refactor to do this in a 2 step manner, first find repo names then fetch with the same method as we do in the other case, but maybe selfishly I want to say I'd rather not make this change for this stupid little optimization bigger?
I'll think a bit if we gain/lose something by trying to unify these 2 paths.

lorenabalan commented on connectors/sources/github.py:
```diff
@@ -1371,50 +1528,50 @@ async def _get_repo_object_for_github_app(self, owner, repo_name):
+                all_repos = configured_repos + foreign_repos
+
+            invalid_repos = []
+            if all_repos:
```
Do we need the if? If it's empty won't `self.github_client.get_repos_by_fully_qualified_name_batch` return empty dict?

Apmats replied:
Indeed. Nice, thanks.

lorenabalan commented on connectors/sources/github.py:
```diff
@@ -1371,50 +1528,50 @@ async def _get_repo_object_for_github_app(self, owner, repo_name):
+                                if org_name not in self.org_repos:
+                                    self.org_repos[org_name] = {}
+                                self.org_repos[org_name][repo_name] = repo_data
+                    else:
+                        self._logger.debug(f"Detected invalid repository: {repo_name}.")
+                        invalid_repos.append(repo_name)
```
Can we reduce the nesting by one level by capturing this at the top of the loop

```python
if not repo_data:
    invalid_repos.appen()
    logger.debug...
    continue
```

I like avoiding messing with the loop control flow but I think this is one of the cases where it's welcome, to make the rest of the code easier to reason about.

lorenabalan replied:
Also, optional, these type of snippets
```python
if org_name not in self.org_repos:
    self.org_repos[org_name] = {}
self.org_repos[org_name][repo_name] = repo_data
```

can become

```python
self.org_repos.setdefault(org_name, {})[repo_name] = repo_data
```

but I don't have a strong opinion if you think the current version is more readable.

lorenabalan reviewed (approved):
LGTM, only small cosmetic suggestions. Thanks for adding typehints along the way! 🙌

Apmats commented:
@artem-shelkovnikov @lorenabalan Apologies - I'll need another approval post-autoformat (my bad on this, still not used to this) and post-merge.

lorenabalan reviewed (approved):
LGTM!
